### PR TITLE
tmz 3.1.0 feature checkdigits

### PR DESCRIPTION
**Pull request by t-m-z**
- IMEI
- IBAN
- Euro
- DEU Tax ID
- EAN/GTIN
- Credit Cards
- ISBN

### CODE REVIEW

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits__datatypes.dart [hunk 1]
```diff
@@ -0,0 +1,82 @@
+part of 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits.dart';
+
+final Map<String, String> ID_LETTERCODE = {
```
Comment (Mike-3) [inline lib/tools/science_and_technology/checkdigits/logic/checkdigits__datatypes.dart:3]: as const if possible

File: lib/application/searchstrings/assets/common.json [hunk 2]
```diff
@@ -82,6 +82,7 @@
   "ccitt_ccir_476": "ccir476 amtor sitor navtex",
   "changelog": "changelog",
   "chao": "chao john francis byrne",
+  "checkdigits": "imei, isbn, ean, gtin, uic, euro, iban, 13616, visacard, mastercard, americanexpress, dinersclub",
```
Comment (Mike-3) [inline lib/application/searchstrings/assets/common.json:85]: remove comma

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+// https://www.bindb.com/bin-database
+// https://www.bindb.com/identify-prepaid-cards
+// https://www.bindb.com/card-schemes
+// https://kreditkarte.net/ratgeber/kreditkartennummer/
+
+// https://www.vccgenerator.org/de/credit-card-validator-result/
+
+// http://www.pruefziffernberechnung.de/K/Kreditkarten.shtml
+
+final Map<String, String> CHECKDIGITS_CREDITCARD_MMI = {
```
Comment (Mike-3) [inline lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart:12]: to const

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+  '3': 'checkdigits_creditcard_major_industry_identifier_3',
+  '4': 'checkdigits_creditcard_major_industry_identifier_4',
+  '5': 'checkdigits_creditcard_major_industry_identifier_5',
+  '6': 'checkdigits_creditcard_major_industry_identifier_6',
+  '7': 'checkdigits_creditcard_major_industry_identifier_7',
+  '8': 'checkdigits_creditcard_major_industry_identifier_8',
+  '9': 'checkdigits_creditcard_major_industry_identifier_9',
+};
+
+final Map<String, String> CHECKDIGITS_CREDITCARD_PREFIX = {
```
Comment (Mike-3) [inline lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart:25]: to const

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+  '69': 'Merchandising-Cards (International)',
+  '81': 'China Union Pay',
+  '90': 'Arca, etc.',
+  '91': 'BelCart, etc.',
+  '94': 'BC Card, Elcart, etc.',
+  '97': 'Troy, ICA, etc.',
+  '98': 'DinaCard, NSMEP, etc.',
+};
+
+final Map<int, List<Object>> CHECKDIGITS_CREDITCARD_CARDSCHEMES = {
```
Comment (Mike-3) [inline lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart:76]: to const

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_ean.dart';
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart';
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_de_taxid.dart';
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_iban.dart';
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_imei.dart';
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_isbn.dart';
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_euro.dart';
+part 'package:gc_wizard/tools/science_and_technology/checkdigits/logic/checkdigits_uic.dart';
+
+final _ITERATE_ALPHA = [
```
Comment (Mike-3) [inline lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart:22]: to const

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+  'S',
+  'T',
+  'U',
+  'V',
+  'W',
+  'X',
+  'Y',
+  'Z',
+];
+final _ITERATE_ALPHANUMERIC = [
```
Comment (Mike-3) [inline lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart:50]: to const
....

File: lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart
Comment (Mike-3) [inline lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart:1]: grafik
Check Credit Card Number
Adjust column widths if possible.

Comment (Mike-3) [reply lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart:1]: Check Credit Card Number
Get Bank/Card Details

No Feedback. 
grafik
Or Percent feedback possible ?

We need Internetis not recognizable that way  ?
If so, please ask whether it is ok to integrate or expand the button text.
Maybe something like geohashing.
If so, please integrate a query as to whether it is ok.
Maybe something like geohashing.

Comment (Mike-3) [reply lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart:1]: Calculate Credit Card Number
grafik
no Output ??

Comment (Mike-3) [reply lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart:1]: Check Credit Card number
grafik
Input: ? and CreditCardnumber (from clipboard) ?1234 1234

Comment (Mike-3) [reply lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart:1]: grafik
 Output also with blocks XXXX XXXX XXXX XXXX

**Review by Mike-3 — CHANGES_REQUESTED** [review]

**Comment by S-Man42** [comment]
I just had a quick view to the UIC tests. But here I cannot see any cases with too long or too short numbers. I guess, there are no such tests elsewhere. Maybe you should focus a bit on edge and invalid cases :)

Could you do me a favor and add 91 80 6493 296-1 to the tests? :) It's my favorite engine ;)

File: lib/application/category_views/all_tools_view.dart [hunk 8]
```diff
@@ -274,13 +274,15 @@ import 'package:gc_wizard/utils/constants.dart';
 import 'package:gc_wizard/utils/string_utils.dart';
 import 'package:gc_wizard/utils/ui_dependent_utils/common_widget_utils.dart';
 import 'package:prefs/prefs.dart';
+import 'selector_lists/checkdigits/checkdigits_selection.dart';
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/application/category_views/all_tools_view.dart:277]: sorry - I do not understand this hint

File: lib/application/searchstrings/assets/common.json (hunk 2 above)
Comment (t-m-z) [reply lib/application/searchstrings/assets/common.json:85]: done

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits__datatypes.dart (hunk 1 above)
Comment (t-m-z) [reply lib/tools/science_and_technology/checkdigits/logic/checkdigits__datatypes.dart:3]: done

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart (hunk 6 above)
Comment (t-m-z) [reply lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart:22]: done

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart (hunk 7 above)
Comment (t-m-z) [reply lib/tools/science_and_technology/checkdigits/logic/checkdigits.dart:50]: done

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart (hunk 4 above)
Comment (t-m-z) [reply lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart:25]: done

File: lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart (hunk 3 above)
Comment (t-m-z) [reply lib/tools/science_and_technology/checkdigits/logic/checkdigits_creditcard.dart:12]: done

File: lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart
Comment (t-m-z) [reply lib/tools/science_and_technology/checkdigits/widget/base/checkdigits_calculate_checkdigit.dart:1]: 1. Check Credit Card Number - Adjust column widths if possible.
   its a lucky game - I did some tests and hope this fits better
   
2. Check Credit Card Number - Get Bank/Card Details - No Feedback.
   - Question/Dialog added
   - percentage not possoble - I do not know what is 100 percent
   
3. Calculate Credit Card Number - No output
   if there is no question marke there is nothing to calculate
   => now there is a check and a hint
   
4. Check Credit Card number - Input: ? and CreditCardnumber (from clipboard) ?1234 1234 - Format exception
   check and corresponding hint added
   
5. Calculate Check digit Output also with blocks XXXX XXXX XXXX XXXX
    format-method added

File: lib/application/category_views/all_tools_view.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/application/category_views/all_tools_view.dart:279]: found the tomatoes and removed them.

[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/application/category_views/all_tools_view.dart]: changes are done within the branch. The master ist damaged. I made a pull request

**Comment by S-Man42** [comment]
grafik

I find it strange, that everythings starts with capital letter except "card". Please make it capital too

**Comment by S-Man42** [comment]
Please add used APIs into the APIs sections of the licenses (at least the bank api): 
grafik

And could you please rename the section to "Used APIs" in that doing?

**Comment by S-Man42** [comment]
grafik

That looks strange from my point of view. Is it meant this way? What should a user read from this?

**Comment by S-Man42** [comment]
978-3-7529-6734-0

How do I enter this IBAN?

**Comment by S-Man42** [comment]
grafik

The line break in the numbers are not necessary, I guess. I believe it comes from too small columns which result that you added the third column for the country, right? Maybe it is the better choice to but DE and Deutschland together into one column (maybe "DE (Deutschland)" or just leave the country code because you can read it from the IBAN itself?) to do not need a break of the numbers

**Comment by S-Man42** [comment]
grafik
 "digit" capital letter, see above. 

Generally: The pattern in the app is: Titles are capitalized, descriptions are normal English writing, means nearly always small letters.

**Comment by S-Man42** [comment]
grafik

Why is here a line break?
Isnt it interesting to see the reason how you identify the printing organisation? (i dont know, is there a code for that, maybe the RB?)

**Comment by t-m-z** [comment]
That's not an iban but isbn

Sent from MailDroid

-----Original Message-----

**Comment by S-Man42** [comment]
I already saw my mistake and deleted the question ;)

**Comment by t-m-z** [comment]
refactor of widget is donw, requested changes are done

**Comment by S-Man42** [comment]
For me still cryptic hint texts (nightly build)
Screenshot_20240228_070221

**Comment by S-Man42** [comment]
Could you please adjust search strings. Was not able to find the Euro tool with "money" maybe there are some good additional strings for the tools :)

Moreover: Are you quite sure about "Euro bill"? I cant find "bill" in any dictionary for Banknote or Geldschein. It is banknote as far as I can see.

**Pull request merged**